Sources/SKCore/BuildSystem.swift

```diff
@@ -62,6 +74,8 @@ public protocol BuildSystem: AnyObject {
 
   /// Called when files in the project change.
   func filesDidChange(_ events: [FileEvent])
+
+  func fileHandlingCapability(for uri: DocumentURI) -> FileHandlingCapability
```

Comment:
If we had async/await, I would say yes. 

Given that we don’t use async/await in SourceKit-LSP yet, I would like to avoid making it async. If `fileHandlingCapability` becomes async, `workspaceForDocument` would also need to be async and thus all request handling functions that call `workspaceForDocument` need to we wrapped in a completion handler – it’s not impossible but does uglify the code. 

Which performance part are you particularly concerned about? AFAICT the most expensive parts of the implementations are and none of them strike me as worrisome enough to justify having to take care of the completion handler in `SourceKitServer`.
- `BuildServerBuildSystem`: Resolve symlinks + a little substring matching
- `BuildSystemManager`: Just calls `max`
- `CompilationDatabaseBuildSystem`: Load compilation database, but should only happen once
- `FallbackBuildSystem`: Always returns `.fallback`
- `SwiftPMWorkspace`: Resolve symlinks + dictionary lookup

Reply:
Okay, let's keep it sync unless it becomes an issue in practice.  

> BuildServerBuildSystem: Resolve symlinks + a little substring matching

This one I'm not sure is correct without querying the server.